### DISCUSSION TITLE
using next-i18next with Next.js 13

### DISCUSSION THREAD

**Discussion by VugarAhmadov** [i18next/next-i18next#1993] [Q&A] [answered]
How we are going to use this package with Nextjs version 13 and how it will work with server components? Because I want my hole page use translation in server and get translated page from server.

previously, we were using this package like this.
`export const getServerSideProps: GetServerSideProps = async ({ locale }) => {`
`return {`
`   props: {`
`      ...(await serverSideTranslations(locale as string, ["common", "login", "form"])),`
`    },`
`  };`
`};`

**Comment by adrai** [comment]
... is there an issue with that?
Check the second page of the examples/simple...
should work as usual

**Comment by VugarAhmadov** [comment]
the example is using pages directory. we can not use getServerSideProps inside app directory. so what will be the alternative?

**Reply by rakibul-wdp** [reply]
@VugarAhmadov Did you solve this issue?
https://github.com/i18next/next-13-app-dir-i18next-example/issues/21#issue-1790796411

**Comment by Gbuomprisco** [comment]
WIth the new `app` directory structure, we will not be able to use `getServerSideProps` and `_app.tsx`. 

There are quite a few changes in there. While next-i18next works great with the `pages` structure, it will not work with the new one.

More info: https://beta.nextjs.org/docs/upgrade-guide#

**Comment by adrai** [comment]
https://beta.nextjs.org/docs/upgrade-guide#migrating-from-pages-to-app
<img width="727" alt="image" src="https://user-images.githubusercontent.com/1086194/197967169-5bbf5700-3550-413c-909f-19797a0ff8db.png">

Feel free to start a PR addressing this.

**Reply by adrai** [reply]
https://github.com/i18next/next-i18next/discussions/1993#discussioncomment-4325016

**Comment by raymonddaikon** [comment]
Also with the new app directory approach `AppProps` is no longer passed to children (since _app and _document don't exist anymore), therefore the `appWithTranslation` HOC does not work.

**Comment by adrai** [comment]
Like said, feel free to start a pull request and contribute to this open source project...

**Reply by adrai** [reply]
https://github.com/i18next/next-i18next/discussions/1993#discussioncomment-4325016

**Comment by Gbuomprisco** [comment]
> Like said, feel free to start a pull request and contribute to this open source project...

I get the why of your comments, "just make a PR" (I do it too!) - but in this case, it's a fundamental change in how this library works with the future Next.js, and I think it would be more appropriate if it were led by a maintainer. 

Please don't take this as "please work for free for me"; it's the beginning of a discussion for making this library work with what Next.js will become, not a command to work on it.

**Comment by adrai** [comment]
That's why I said "...feel free to **start** a pull request..." ;-)

**Comment by Gbuomprisco** [comment]
Before I start a PR, should this be moved to a discussion, maybe? Since the API will have significant changes, and maintainers should lead the design of those?

**Comment by LB22** [comment]
https://beta.nextjs.org/docs/app-directory-roadmap
> **Internationalization (i18n)**
We will be providing more guidance on solutions for internationalized routing inside app. With the addition of [generateStaticParams](https://beta.nextjs.org/docs/api-reference/generate-static-params), which can be used inside layouts, this will provide improved flexibility for handling routing when paired with [Middleware](https://nextjs.org/docs/advanced-features/middleware) versus the existing i18n routing features in pages/.

**Reply by jlarmstrongiv** [reply]
More guidance can be found here https://github.com/vercel/next.js/issues/41980#issuecomment-1328311472

**Comment by dejurin** [comment]
You can use native 

```
'use client'

import i18n from "i18next";
import { useTranslation, initReactI18next } from "react-i18next";


i18n
.use(initReactI18next) // passes i18n down to react-i18next
.init({
  // the translations
  // (tip move them in a JSON file and import them,
  // or even better, manage them via a UI: https://react.i18next.com/guides/multiple-translation-files#manage-your-translations-with-a-management-gui)
  resources: {
    en: {
      translation: {
        "Welcome to React": "Welcome to React and react-i18next"
      }
    }
  },
  lng: "en", // if you're using a language detector, do not define the lng option
  fallbackLng: "en",

  interpolation: {
    escapeValue: false // react already safes from xss => https://www.i18next.com/translation-function/interpolation#unescape
  }
});


function RootLayout({
  children,
}: {
  children: React.ReactNode;
}): JSX.Element {

  const { t } = useTranslation();


  
  return (
    <html lang="en">
      <head>
      </head>
      <body>
            <h2>{t('Welcome to React')}</h2>
            {children}
      </body>
    </html>
  );
}

export default RootLayout;
```

**Reply by adrai** [reply]
This will work directly only in the client, right?

**Reply by dejurin** [reply]
Yep, I was use 
**'use client'**

**Reply by aldabil21** [reply]
with this setup, you will use client components everywhere you use the `t` function, which is almost everywhere. Thats not what Next 13 is all about (Server component first and by default).

**Reply by adrai** [reply]
https://github.com/i18next/next-i18next/discussions/1993#discussioncomment-4325016

**Comment by murad-alqadi** [comment]
What is the current consensus on roadmap for addressing this issue?

**Reply by adrai** [reply]
https://github.com/i18next/next-i18next/discussions/1993#discussioncomment-3968268

**Comment by adrai** [comment]
> What is the current consensus on roadmap for addressing this issue?

https://github.com/i18next/next-i18next/discussions/1993#discussioncomment-3968268

**Comment by adrai** [comment]
fyi: https://beta.nextjs.org/docs/app-directory-roadmap#not-planned-features
<img width="744" alt="image" src="https://user-images.githubusercontent.com/1086194/205887621-77b089ad-510a-43bc-9fb7-15ff14dc807c.png">

**Reply by thulstrup** [reply]
Vercel did provide a very simple example of using internationalization in the `app` directory.
https://github.com/vercel/examples/tree/main/app-directory/i18n

**Reply by adrai** [reply]
There's still no support for i18n... that's just a workaround... and that's also doable with i18next directly... no react-i18next and no next-i18next is needed for that.

**Reply by adrai** [reply]
Will create a little example how this could work directly with i18next with the same "workaround"...

**Reply by LB22** [reply]
If you @adrai, or someone else who has strong knowledge and ties to i18next / react-i18next / next-i18next, would get in contact with the Next.js team to understand their vision, get advice for next-i18next, manage pain points, open up a feedback channel, etc, I think that could help this package out tremendously. I'm sure they are aware of this package.

With "not planning to include the i18n feature", I think they mean that they won't specifically build an i18n solution such as next-i18next into Next.js by default. I'm sure they however want something like next-i18next to exist, and I'm sure they are willing to help out keeping their own ecosystem (next-i18next) alive.

Again, someone with strong knowledge of this package should get in contact with them and hopefully they could offer us something valuable.

https://github.com/vercel/next.js/issues/41980#issuecomment-1328311472
"Open to any feedback" - They want us to give our opinions

**Reply by adrai** [reply]
Here a simple example that uses i18next and react-i18next directly... without the need of next-i18next: https://github.com/i18next/next-13-app-dir-i18next-example

**Reply by dejurin** [reply]
In ur example, have a two dir with json files. Why u keep "public"? This example not used public's json files, yep?
Sorry, if this noob question, I'm beginner.

**Reply by adrai** [reply]
Sorry, that is old stuff: https://github.com/i18next/next-13-app-dir-i18next-example/commit/6108fa0c31aa292be785175865c8f7e7169b104d

**Comment by dejurin** [comment]
initI18nextInstance.tsx
**UPD**
```
import i18next from 'i18next';
import resourcesToBackend from 'i18next-resources-to-backend';
// import { initReactI18next } from 'react-i18next'; // <-- wrong
import { initReactI18next } from 'react-i18next/initReactI18next'; // <-- OK

export const fallbackLng = 'en';
export const languages = [fallbackLng];
export const defaultNS = 'common';

const initI18next = async (lng: string, ns: string) => {
  // if (ns && ns.indexOf(defaultNS) < 0) ns.push(defaultNS)
  const i18nInstance = i18next.createInstance();
  await i18nInstance
    .use(initReactI18next)
    .use(resourcesToBackend((language, namespace, callback) => {
      import(`./locales/${language}/${namespace}.json`)
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        .then((resources) => callback(null, resources))
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        .catch((error) => callback(error, null));
    }))
    .init({
      debug: true,
      initImmediate: true,
      supportedLngs: languages,
      // preload: languages,
      fallbackLng,
      lng,
      fallbackNS: defaultNS,
      defaultNS,
      ns,
      interpolation: {
        escapeValue: false,
      },
    });
  return i18nInstance;
};

export default async function initI18nextInstance(lng: string, ns: string, keyPrefix?: undefined) {
  const i18nextInstance = await initI18next(lng, ns);
  return {
    t: i18nextInstance.getFixedT(lng, ns, keyPrefix),
    i18n: i18nextInstance,
  };
}

```
page.tsx
```
import initI18nextInstance from '../../i18n/initI18nextInstance';

export default async function Page({ params }: { params: { lng: string } }) {
  const { lng } = params;
  const { t } = await initI18nextInstance('en', 'common');
  return <>
    {t('test')}
  </>;
}
```

I got Error, when I'm using **initReactI18next** 
<img width="1005" alt="Screenshot 2022-12-07 at 16 33 39" src="https://user-images.githubusercontent.com/4470015/206207578-9a3eb990-7b4b-42af-9749-685f1c8282e2.png">

if I delete, everythin is ok... What I'm doing wrong?

**Reply by adrai** [reply]
You're probably importing react-i18next in a wrong way somewhere...
Create a minimal reproducible example so we can check.

**Reply by dejurin** [reply]
OMG, 
`import { initReactI18next } from 'react-i18next'; `
))))) this mistake

**Comment by adrai** [comment]
The example here: https://github.com/i18next/next-13-app-dir-i18next-example is showing how i18next could be used on server side and client side... and also a middleware with language detection and cookie persistence...
And here the corresponding blog post: https://locize.com/blog/next-13-app-dir-i18n/

**Reply by Baraboo88** [reply]
Hello @adrai, hope you are doing well. 
Thank you  for the great guide.
Tried to implement the above example using typescript and see the below error.


Do you have any thoughts how to fit it?

![image](https://user-images.githubusercontent.com/42747313/208368515-ac20c480-1140-48c5-99af-6030b14fcea5.png)

Thanks and best regards

**Reply by adrai** [reply]
I'm not a TypeScript expert, but: https://github.com/i18next/next-13-app-dir-i18next-example-ts
// cc: @pedrodurek

**Reply by scriptsman** [reply]
> The example here: https://github.com/i18next/next-13-app-dir-i18next-example is showing how i18next could be used on server side and client side... and also a middleware with language detection and cookie persistence... And here the corresponding blog post: https://locize.com/blog/next-13-app-dir-i18n/

Thanks for the example. It seems that we must pass the lang to every server component in props. 

Is there any way to avoid this?

**Reply by adrai** [reply]
> Is there any way to avoid this?

@scriptsman I'm not aware of any way to avoid it right now.

**Reply by aureliopita-au** [reply]
I'm also interested in a way of using `useTranslation()` without the need to send the language.

**Reply by christianvuerings** [reply]
We set the language as a header in Next.js middleware and then use it in the server version of `useTranslation`:

```tsx
import { headers } from "next/headers";
//
export async function useTranslation<N extends Namespace>(ns?: N) {
  const headersList = headers();
  const lang = headersList.get("x-lang") ?? fallbackLng;
```

**Reply by adrai** [reply]
https://github.com/i18next/next-13-app-dir-i18next-example/issues/12

**Reply by weilin-ong** [reply]
Hey thanks for the example. Quick question, can both setups co-exist?
I'm slowly migrating my app from pages router to app router, while every file works fine under the pages router, the migrated files under app router do not. Can I not delete the old setup while I'm following this new approach? https://locize.com/blog/next-13-app-dir-i18n/

**Reply by adrai** [reply]
What do you mean by "the migrated files under app router do not"? Is there an error?
Never tested that myself...

**Reply by weilin-ong** [reply]
yes, i got this error: 
--> react-i18next:: You will need to pass in an i18next instance by using initReactI18next 

i think the error comes from removing this part in my migrated files, i had to remove it because i cant use getStaticProps in client components:
```
import { serverSideTranslations } from 'next-i18next/serverSideTranslations'
export async function getStaticProps({ locale }) {
  return {
    props: {
      ...(await serverSideTranslations(locale, [
        'common',
        'footer',
      ])),
      // Will be passed to the page component as props
    },
  }
}
```

**Reply by adrai** [reply]
in the app router part you can't use next-i18next anymore, but in the pages part you need to import useTranslation from next-i18next

**Comment by paulghc051080** [comment]
Do you have an example of how to use dynamic routes? I had an error when I created the dynamic page folder 'Server Error
Error: A required parameter (name) was not provided as a string in generateStaticParams for /[lng]/[name]

This error happened while generating the page. Any console logs will be displayed in the terminal window.' thanks

**Reply by adrai** [reply]
app directory does not have the concept for dynamic routes, but dynamic segments, if you have next.js specific questions you might ask directly in the vercel/next.js repo

**Comment by albertothedev** [comment]
Does this mean that this library is useless since Next.js 13 and we should instead use `i18next` and `react-i18next`? From the blog post:

> Nice features provided by [next-i18next](https://next.i18next.com/) (and other Next.js related i18n modules), like described [here](https://locize.com/blog/next-i18next/) and [here](https://locize.com/blog/next-i18n-static/) are not suited to this new app directory setup.

**Reply by adrai** [reply]
It just means next-i18next is not necessary anymore:
https://locize.com/blog/next-13-app-dir-i18n/
https://github.com/i18next/next-13-app-dir-i18next-example

**Reply by albertothedev** [reply]
Please, correct me if I get anything wrong. 

So this library is useless for projects using the new `app` directory. In fact, it's not even compatible with it and there are no plans to make it so. People starting a project with the new API or upgrading their current one to it should instead use `i18next` and `react-i18next` like described in the blog post and the example. The only use case for this library is for people using the `pages` directory.

Shouldn't that be one of  the first things a user sees when visiting the repo and the NPM page? Next.js 13 released over 6 months ago, many people have switched or are in the process of doing so. The only reference to the need for this new setup is on the website:

> Are you using the new Next.js 13 app directory?
>> Then check out [this blog post](https://locize.com/blog/next-i18next/).

and neither the website nor the blog post properly convey its importance. I think having a disclaimer like 

> This library isn't compatible with Next.js 13's `app` directory. If you are using it, there's a new setup using [i18next](https://www.i18next.com/) and [react-i18next](https://react.i18next.com/) explained [here](https://locize.com/blog/next-13-app-dir-i18n/)

as the first line in the repo's readme and the NPM package page to avoid confusion, which is exacerbated by this library's release versioning being so similar to Next.js's.

**Reply by adrai** [reply]
next-i18next is still compatible with Next.js 13... but only for the pages setup, and not with the app directory setup.
<img width="867" alt="image" src="https://user-images.githubusercontent.com/1086194/235244240-1417fea8-01b9-4926-910b-5eaa814123f9.png">

**Reply by albertothedev** [reply]
Great, I saw the commits while I was working on my PR #2148. I think this works.

**Reply by sagargulati** [reply]
So next-i18next doesn't seem to be compatible with the new app directory at all and it doesn't seem to be on roadmap too?

**Reply by adrai** [reply]
like said, it's just not necessary anymore.

**Reply by albertothedev** [reply]
It would be great to not have to manually configure the middleware.

**Comment by Ivan275** [comment]
Elint complaint this when I'm building: Error: React Hook "useTranslation" cannot be called in an async function.  react-hooks/rules-of-hooks

**Reply by aldabil21** [reply]
Rename it, for example rename the server-side `useTranslation` to `getTranslation`

**Reply by Ivan275** [reply]
Thanks!, that works!

**Comment by multivoltage** [comment]
Hi.
I have a scenario probably equals to most of you.
I developed my entire site with next 12 and used next-i18-next. 
Today my site use next 13 and only 2 new pages uses app router. 

This because all my pages have got `footer` and `header` component which are using `useTranslation` from `next-i18-next`.

these 2 pages fortunatly ara pegas without any component (only a canvas).

But now when I need to create a new page I would create in the app router. But this is impossible since a normal page in my case has both `footer` and `header` component. They (with `use client` on top) did not work since requires code like `...(await serverSideTranslations(locale, ["common"])),`

How can I start a migration? I dupicate `footer` and make it working with app router?

**Comment by adrai** [comment, marked as answer]
**Update (2026): next-i18next v16 supports the App Router.**

This thread and its older answers predate v16. Since next-i18next v16.0.0 (March 2026), the App Router is supported natively, alongside the Pages Router and mixed setups:

- `getT()` for Server Components
- `useT()` for Client Components
- `createProxy()` for language detection and routing (`proxy.ts` / `middleware.ts`)

Docs: https://github.com/i18next/next-i18next#readme
Runnable examples (locale-in-path, no-locale-path, mixed routers): https://github.com/i18next/next-i18next/tree/master/examples
Background on the change: https://www.locize.com/blog/next-i18next-v16

The earlier comments here recommending "use i18next / react-i18next directly, next-i18next is not needed for the app directory" reflect the pre-v16 situation. That direct setup still works (see https://github.com/i18next/next-app-dir-i18next-example), but next-i18next is now the batteries-included path for the App Router.
